himbaechel/uarch/gowin/gowin_arch_gen.py

```diff
@@ -0,0 +1,701 @@
+            direction = uturnlut[direction]
+        if x > db.cols - 1:
+            x = 2 * db.cols - 1 - x
+            direction = uturnlut[direction]
+        wire = f'{direction}{num}{segment}'
+    return (x, y, wire)
+
+def create_nodes(chip: Chip, db: chipdb):
+    # : (x, y)
+    dirs = { 'N': (0, -1), 'S': (0, 1), 'W': (-1, 0), 'E': (1, 0) }
+    X = db.cols
+    Y = db.rows
+    global_nodes = {}
+    for y in range(Y):
+        for x in range(X):
+            nodes = []
+            tt = chip.tile_type_at(x, y)
+            extra_tile_data = tt.extra_data
+            # SN and EW
+            for i in [1, 2]:
+                nodes.append([NodeWire(x, y, f'SN{i}0'),
+                        NodeWire(*uturn(db, x, y - 1, f'N1{i}1')),
+                        NodeWire(*uturn(db, x, y + 1, f'S1{i}1'))])
+                nodes.append([NodeWire(x, y, f'EW{i}0'),
+                        NodeWire(*uturn(db, x - 1, y, f'W1{i}1')),
+                        NodeWire(*uturn(db, x + 1, y, f'E1{i}1'))])
+            for d, offs in dirs.items():
+                # 1-hop
+                for i in [0, 3]:
+                    nodes.append([NodeWire(x, y, f'{d}1{i}0'),
+                        NodeWire(*uturn(db, x + offs[0], y + offs[1], f'{d}1{i}1'))])
+                # 2-hop
+                for i in range(8):
+                    nodes.append([NodeWire(x, y, f'{d}2{i}0'),
+                        NodeWire(*uturn(db, x + offs[0], y + offs[1], f'{d}2{i}1')),
+                        NodeWire(*uturn(db, x + offs[0] * 2, y + offs[1] * 2, f'{d}2{i}2'))])
+                # 4-hop
+                for i in range(4):
+                    nodes.append([NodeWire(x, y, f'{d}8{i}0'),
+                        NodeWire(*uturn(db, x + offs[0] * 4, y + offs[1] * 4, f'{d}8{i}4')),
+                        NodeWire(*uturn(db, x + offs[0] * 8, y + offs[1] * 8, f'{d}8{i}8'))])
+            # I0 for MUX2_LUT8
+            if (x < X - 1 and extra_tile_data.tile_class == chip.strs.id('LOGIC')
+                    and  chip.tile_type_at(x + 1, y).extra_data.tile_class == chip.strs.id('LOGIC')):
+                nodes.append([NodeWire(x, y, 'OF30'),
+                             NodeWire(x + 1, y, 'OF3')])
+
+            # ALU
+            if extra_tile_data.tile_class == chip.strs.id('LOGIC'):
+                # local carry chain
+                for i in range(5):
+                    nodes.append([NodeWire(x, y, f'COUT{i}'),
+                                  NodeWire(x, y, f'CIN{i + 1}')]);
+                # gobal carry chain
+                if x > 1 and chip.tile_type_at(x - 1, y).extra_data.tile_class == chip.strs.id('LOGIC'):
+                    nodes.append([NodeWire(x, y, f'CIN0'),
+                                  NodeWire(x - 1, y, f'COUT5')])
+
+            for node in nodes:
+                chip.add_node(node)
+
+            # VCC and VSS sources in the all tiles
+            global_nodes.setdefault('GND', []).append(NodeWire(x, y, 'VSS'))
```

Comment:
as a heads up to myself - having a node that fans out to all tiles will cause problems with current deduplication. you don't need to change anything right now, but beware the database will be bigger than expected for this reason (the same is possibly true for global nodes, I will need to see how well the apicula structure compacts)

Reply:
Fixed.

Reply:
Base sizes are still acceptable :)

![bases](https://github.com/YosysHQ/nextpnr/assets/6075465/49f9c9ce-71f3-4541-9b03-34545ee9155b)

Reply:
That's really nice - I suspect this is probably only a big problem when talking about the millions of LUTs then ^^

Will put this as more of a slightly background thing compared to the other TODOs on my side.